### PR TITLE
Implement support for multiple args for repo remove

### PR DESCRIPTION
**Pull request by lindhe**
I found that I sometimes I want to remove multiple repos at once.
Thought I would give it a Go to try and to make that a bit easier to do.

This is my first contribution to the Helm code base, and I am doing this
party to become more familiar with Helm and Go. So please point me in
the right direction if I've missed to change something.

Usually, I'd create an issue before creating a PR. If this change is
unwanted, please feel free to discard it – no hard feelings. I made it
mostly as an exercise anyway, so it was not wasted time for me.

### CODE REVIEW

**Comment by lindhe** [comment]
Ah, tests. Cool. I'll try to go through them at some opportune moment.

File: cmd/helm/repo_remove.go
Comment (marckhouzam) [inline cmd/helm/repo_remove.go:46]: Please update the Short description for this new behavior

**Review by marckhouzam — CHANGES_REQUESTED** [review]
I like this. It gives users a little more flexibility.

Can you update the completion code a little lower to match? It currently only provides repo names as completions if there are no other arguments. With your change it should provide repo names all the time.  If you want to be really thorough, the completions could ignore any repo already provided by previous arguments.

File: cmd/helm/repo_remove.go
Comment (lindhe) [reply cmd/helm/repo_remove.go:46]: Got it. :) Will look into it.

**Comment by lindhe** [comment]
It wasn't trivial for me to make it ignore already completed repos, but I'll give it one more shot.

**Comment by marckhouzam** [comment]
> It wasn't trivial for me to make it ignore already completed repos, but I'll give it one more shot.

The list of repos passed in is in the `args` parameter of the completion function. You can probably just loop over the `args` and remove the existing ones from the received completion list.  It's basically removing all elements of one list from another list. 

Then again, this is just completion so it's not a big deal if we suggest a repo that's already being removed. The user just won't select it.  The cool thing about filtering is that a user could know what repos are left by simply typing TAB as she adds repos to the removal command.

It's not a blocking request.

**Comment by marckhouzam** [comment]
@lindhe your last commit is not signed. You should force-push an updated one please.

**Comment by lindhe** [comment]
What about this solution? Any good? It looks like it has quadratic complexity, but perhaps it's not a big deal in this context?

File: cmd/helm/repo_list.go
Comment (marckhouzam) [inline cmd/helm/repo_list.go:110]: Since we know that `ignored` will always be of size `len(ignoredRepoNames)`, how about setting that size in the call to `make`?

File: cmd/helm/repo_list.go
Comment (marckhouzam) [inline cmd/helm/repo_list.go:111]: Nit: to avoid using the same variable name with two different types, can you rename `repo` here to `name` or `repoName`?

**Review by marckhouzam — CHANGES_REQUESTED** [review]
Nicely done! Two little nits and we're good to go.

**Comment by marckhouzam** [comment]
Oh and the `repo_remove_test.go` has to be updated to match, as it is failing now.
You can do
```
make test>result
```
You will see it no longer compiles

**Comment by marckhouzam** [comment]
One other request @lindhe, if you don't mind.  Can you add a test that removes more than one repo and checks that they are both removed?

**Comment by lindhe** [comment]
> One other request @lindhe, if you don't mind. Can you add a test that removes more than one repo and checks that they are both removed?

I very much agree that we should test that, but I don't have more time to work on this today. Feel free to write such tests, or we can just wait until a day when I have more time. :) Not sure if that'll be tomorrow or in two weeks though.

**Comment by lindhe** [comment]
BTW, it seems I'm missing at least `golangci-lint` because `make test` fails and tells me that. Any pointers on how to install it (on Ubuntu)?

**Comment by marckhouzam** [comment]
> BTW, it seems I'm missing at least `golangci-lint` because `make test` fails and tells me that. Any pointers on how to install it (on Ubuntu)?
```
curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(go env GOPATH)/bin v1.24.0
```
taken from https://github.com/golangci/golangci-lint#install

I'll let you finish off writing the test, whenever you have time. You've come far enough to deserve full credit for this improvement.

**Comment by lindhe** [comment]
It'd be best to keep all the old tests, right? Just adding a new test that also tests multiple repos at once?

**Comment by marckhouzam** [comment]
Yes, might as well keep the existing test to make sure removing a single repo works, and adding a very similar test for removing two or three repos.

**Comment by lindhe** [comment]
I had to refactor the code a bit to avoid too much copy-paste programming. But I don't quite grasp the parts that I had to refactor (the use of `cacheIndexFile` and `cacheChartFile`), so would definitely be good with some eyes on that.

File: cmd/helm/repo_remove_test.go
Comment (lindhe) [inline cmd/helm/repo_remove_test.go:87]: Is it right to use the same buffer object for all three repos?

File: cmd/helm/repo_remove_test.go
Comment (lindhe) [inline cmd/helm/repo_remove_test.go:129]: I'm pretty unhappy about this solution. Looks pretty messy, and the definition of `idxs` on line 102 is pretty bad too. Do we even need to check each repo individually, or can we perhaps check them all in one swoop, getting rid of having to store the index files for each repo...?

File: cmd/helm/repo_remove_test.go
Comment (marckhouzam) [reply cmd/helm/repo_remove_test.go:87]: The buffer is simply used to gather the printouts of the remove command.  You can use the same buffer throughout the function as long as you call `b.Reset()` before calling a new `repo remove` command.

File: cmd/helm/repo_remove_test.go
Comment (marckhouzam) [reply cmd/helm/repo_remove_test.go:129]: Personally, I would go simpler.  You are creating two repos, and each repo will have an index file and a cache file.  The test just needs to make sure both files are removed after the repos are removed.  I would just have 4 variables. Instead of the old `idx` and `idx2` which were poorly named anyway, I would simply have  `indexFile1`, `cacheFile1` and `indexFile2`, `cacheFile2`.

File: cmd/helm/repo_remove_test.go
Comment (marckhouzam) [inline cmd/helm/repo_remove_test.go:133]: The buffer is not related to the handling below, so you don't need to pass it in.  The `b.Reset()` should just be done before the call to `o.Run(b)` so that the output of the `repo remove` command will end up in an empty buffer.

File: cmd/helm/repo_remove_test.go
Comment (marckhouzam) [inline cmd/helm/repo_remove_test.go:72]: The above code is cryptic.  What it means to do is to get a reference to the indexFile and to the cacheFile, so that after the `repo remove` command , we have a handle to check that both files no longer exist.

**Review by marckhouzam — CHANGES_REQUESTED** [review]
Yeah, the test is a bit cryptic.  I tried to explain a bit.  Let me know if you want more clarifications.

**Review by marckhouzam — APPROVED** [review]

**Comment by marckhouzam** [comment]
LGTM!
Thanks for you perseverance @lindhe 

Because of the tests, the size of the PR is now considered "large", so I need another maintainer to approve also.

File: cmd/helm/repo_list.go
Comment (technosophos) [inline cmd/helm/repo_list.go:116]: nit: This could be reduced to:

```go
if _, removed := ignored[repo.Name]; !removed {...
```

Not a blocker.

**Comment by lindhe** [comment]
I intended to also clean up the flow of the test code (at least the bit from @marckhouzam's [suggestions](https://github.com/helm/helm/pull/7791#discussion_r404506435)), because I'm having a hard time following the current logic.

But I'm not particularly enthusiastic about rewriting some working but messy code, and I'm not sure when I have the time and motivation for it. So if you are happy with the current state, maybe we call this PR done and leave the tidying up for another PR. Your choice.

File: cmd/helm/repo_remove_test.go
Comment (technosophos) [inline cmd/helm/repo_remove_test.go:143]: This function name does not make sense to me. Can we name it something that better indicates that it is just creating two files?

**Review by technosophos — APPROVED** [review]
Finished reviewing the code, and it is good to go. I left one non-blocking nit.

@marckhouzam : Feel free to merge whenever you are ready.

File: cmd/helm/repo_remove_test.go
Comment (marckhouzam) [reply cmd/helm/repo_remove_test.go:143]: Since it was brought up, I agree, might as well give it a descriptive name.
So, one last change please @lindhe and then we merge.

**Comment by lindhe** [comment]
I took the liberty to improve some other names too.

**Comment by marckhouzam** [comment]
Thanks @lindhe, this looks great.

And thanks @technosophos for the second review.

**Pull request merged**
